File: buildSrc/src/main/java/io/opentelemetry/instrumentation/gradle/muzzle/MuzzleExtension.java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.instrumentation.gradle.muzzle;

import javax.inject.Inject;
import org.gradle.api.Action;
import org.gradle.api.model.ObjectFactory;
import org.gradle.api.provider.ListProperty;

public abstract class MuzzleExtension {

  private final ObjectFactory objectFactory;

  @Inject
  public MuzzleExtension(ObjectFactory objectFactory) {
    this.objectFactory = objectFactory;
  }

  public abstract ListProperty<MuzzleDirective> getDirectives();

  public void pass(Action<? super MuzzleDirective> action) {
    MuzzleDirective pass = objectFactory.newInstance(MuzzleDirective.class);
    action.execute(pass);
    pass.getAssertPass().set(true);
    getDirectives().add(pass);
  }

  public void fail(Action<? super MuzzleDirective> action) {
    MuzzleDirective fail = objectFactory.newInstance(MuzzleDirective.class);
    action.execute(fail);
    fail.getAssertPass().set(false);
    getDirectives().add(fail);
  }
}

File: gradle-plugins/src/main/kotlin/io/opentelemetry/javaagent/muzzle/MuzzleExtension.kt
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.javaagent.muzzle

import javax.inject.Inject
import org.gradle.api.Action
import org.gradle.api.model.ObjectFactory
import org.gradle.api.provider.ListProperty

abstract class MuzzleExtension @Inject constructor(private val objectFactory: ObjectFactory) {

  internal abstract val directives: ListProperty<MuzzleDirective>

  fun pass(action: Action<MuzzleDirective>) {
    val pass = objectFactory.newInstance(MuzzleDirective::class.java)
    action.execute(pass)
    pass.assertPass.set(true)
    directives.add(pass)
  }

  fun fail(action: Action<MuzzleDirective>) {
    val fail = objectFactory.newInstance(MuzzleDirective::class.java)
    action.execute(fail)
    fail.assertPass.set(false)
    directives.add(fail)
  }
}
